interval_tree.py

```diff
@@ -0,0 +1,88 @@
+def build_tree(data, reducing_fn):
+    n = len(data)
+    i = next(2**j for j in range(n + 1) if 2**j >= n)
+    tree = [0] * i * 2
+
+    floor = data + [0] * (i - n)
+    while True:
+        if i == 0: return tree
+        tree[i:2 * i] = floor
+        i //= 2
+        floor = [reducing_fn(floor[2 * j:2 * j + 2]) for j in range(i)]
+
+def left_child(i, grandness=1):
+    return i * 2**grandness
+
+def right_child(i, grandness=1):
+    return i * 2**grandness + 2**grandness - 1
+
+def parent(i):
+    return i // 2
+
+# including a, including b
+def resolve_query(tree, reducing_fn, a, b):
+    values = []
+    n = len(tree)
+
+    def add_vals(i, k):
+        start = left_child(i, k) - n // 2
+        end = right_child(i, k) - n // 2
+
+        if a <= start and b >= end:
+            values.append(tree[i])
+        elif (a >= start and a <= end) or (b >= start and b <= end):
+            add_vals(2 * i, k - 1)
+            add_vals(2 * i + 1, k - 1)
+
+    N = next(i for i in range(n) if 2**i == n) - 1
+    add_vals(1, N)
+    return reducing_fn(values)
```

Comment:
I'd like more if reducing_fn takes individual arguments to reduce instead of an array